### ISSUE TITLE
`argo-ui`: future state of this repo

### ISSUE THREAD

**Issue by agilgur5** [argoproj/argo-ui#453]
This is based on discussions in [this Slack thread](https://cloud-native.slack.com/archives/C020XM04CUW/p1692085374857769) (which is fairly long and winding) that I had raised after realizing the fairly neglected and incomplete state of this repo when having to do an upgrade of it for Argo Workflows in https://github.com/argoproj/argo-workflows/pull/11585. 

This was also mentioned in a [Maintainers meeting](https://docs.google.com/document/d/1C71GA0XCxsyGP7Q4SYveHak5ZZcJOuaOiDCt_xHG10U/edit#heading=h.6upblgdtbf14) that I was not part of.

## Future State -- intent to phase out(?)

This repo has been neglected for some time and right now there are not that many cross-project contributors. Due to this and the interim state of this repo as not entirely a library (see below), this has resulted in slowing down development while simultaneously not really creating a "consistent look and feel" between projects (such as Workflows and CD). There are some similarities and shared branding, but many differences (in part because CD has received more attention) despite having some shared components.

While it may be possible to improve this situation, some contributors believe that this may not be worthwhile and that a design guide of sorts may be a better substitute. Argoproj also already uses [`antd`](https://github.com/ant-design/ant-design) as the underlying component library. This repo has higher-level components on top of those as well as some components that are not part of `antd`.
(Personally, I do not really have a strong opinion myself as I am a more recent contributor to Argoproj and am more documenting the current thoughts around this for everyone. I think the "ideal state" of this repo could be great, but if it does not suffice for those purposes and may not ever suffice, while also currently slowing things down, phasing out may certainly make a lot of sense.)

### Requiring downstream issues
In any case, as the repo is indeed not really actively maintained (and has no active maintainer list), we should document its current state.

Instead of having contributors create issues and PRs here ad-hoc, it may be better in the interim to require that issues and PRs have a downstream issue in consuming projects like Argo Workflows or Argo CD. That way, maintainers of the consuming projects will be aware of and can review PRs here.

### TypeScript compatibility issues
I may perhaps be the only one interested in improving this, but as I mentioned in https://github.com/argoproj/argo-workflows/pull/11510#issuecomment-1678520239, this repo currently exports raw TSX, instead of compiled JS. This causes compat issues when consuming projects have different TS versions or settings, which means that the settings and version of this repo largely define the settings and version of consuming repos. It also means longer compilation times and that consuming repos must have many of the same devDeps despite not using them directly (which is _very_ confusing and hard to determine -- package managers will not automatically determine this given that this is not how devDeps are meant to be used).

Of particular note is that this repo is on [TS 4.9](https://github.com/argoproj/argo-ui/blob/76fa7c1ae65503c2c9730eb4fa95402ac05c815c/package.json#L40) and does not have the strictest (or even strict, for that matter) type-checking [configuration](https://github.com/argoproj/argo-ui/blob/76fa7c1ae65503c2c9730eb4fa95402ac05c815c/tsconfig.json).

This makes it all the more harder to decouple this repo from consuming projects as if they use any components from here, they are hard tied to its version and settings. Compiling the TSX to JS and publishing it, whether in the NPM package or directly in the repo, would significantly improve this situation.

## History

This is more of a miscellaneous note, but I think the history of a repo is important context to any decisions.

### Argo Workflows
As far as I know, this repo was initially the UI for Argo Workflows. Then it evolved as the Argoproj org evolved, including Argo CD and others. This can be seen in late-2019 PRs #53 and https://github.com/argoproj/argo-workflows/pull/1859 that move the Workflows UI into Argo Workflows. (Notably, COVID-19 started to make its mark on the entire world around that time).

### Component Library
Then there was some momentum from contributors to make this repo into a component library for all Argo Projects. Can see some mid-2021 PRs such as #96.

This appears to have never quite been completed though, and this repo has been in this interim state ever since.

### v1 vs. v2
I don't know that many details here, but if I'm interpreting correctly, v1 was a first-pass assortment of components and v2 was meant to be ["a true component library which is shared between all Argo Projects and published on NPM"](https://cloud-native.slack.com/archives/C020XM04CUW/p1693246601697549?thread_ts=1692085374.857769&cid=C020XM04CUW).

It looks like it was also supposed to modern hooks syntax and other things. There is not a 1-to-1 component match though.

**Comment by agilgur5** [comment]
@terrytangyuan would you mind pinning this issue?
Sorry to have to keep bugging you for assistance; tbh that's probably the main reason I want write permissions at this point, so I can do certain things without needing to ask approver+ folks 😅

**Comment by agilgur5** [comment]
Jotting down https://github.com/argoproj/argo-ui/pull/432#issuecomment-1879816619 -- upgrading to Storybook v7 and Webpack v5 is highly non-trivial and may not be worthwhile given the state of this repo

**Comment by agilgur5** [comment]
> While it may be possible to improve this situation, some contributors believe that this may not be worthwhile and that a design guide of sorts may be a better substitute. Argoproj also already uses [`antd`](https://github.com/ant-design/ant-design) as the underlying component library. This repo has higher-level components on top of those as well as some components that are not part of `antd`.

Ok apparently `antd` is actually not used by CD or Workflows or even `argo-ui` itself as I learned in https://github.com/argoproj/argo-ui/issues/553#issuecomment-2106048778. The `./antd` directory only has Storybook stories and no actual components and is entirely unused.
As I wrote there, Rollouts does use `antd` _directly_ in various places, and that's probably what CD and Workflows should move to do as well.

**Comment by andrii-korotkov** [comment]
Would there be a new tag for this? The current one is almost 2 years old.

**Comment by andrii-korotkov** [comment]
Yeah, not needed. I already got an answer that commit reference is used by projects, which I didn’t know before.

**Comment by agilgur5** [comment]
> Jotting down [#432 (comment)](https://github.com/argoproj/argo-ui/pull/432#issuecomment-1879816619) -- upgrading to Storybook v7 and Webpack v5 is highly non-trivial and may not be worthwhile given the state of this repo

I tried a small bit more on the upgrade as I found [another automigration](https://storybook.js.org/docs/7/migration-guide#storiesof-to-csf) that I didn't know about before, but still had issues. Threw up a draft of some v1 changes fwiw though 🤷 : #568  

Also tried to downgrade to fix #469 and that failed spectacularly as well: #567
Given the historical details I found there (lack of working Storybook build for _years_ and incompatible changes by humans suggesting it was not used for dev or testing), I am considering nuking the Storybook entirely to continue reducing the scope of this repo in furtherance of this issue and to leave no confusion about that possibly working. That would remove a good bit of code from this repo as well as lots of deps.
